File: 프로그래머스/lv3/64062. 징검다리 건너기/징검다리 건너기.py
def possible(mid, stones, k):
    cnt = 0
    for stone in stones:
        if mid > stone:
            cnt += 1
            if cnt >= k:
                return False
        else:
            cnt = 0
    return True

def solution(stones, k):
    start, end = 0, max(stones)
    ans = 0
    
    while start <= end:
        mid = (start + end) // 2
        if possible(mid, stones, k):
            ans = max(ans, mid)
            start = mid + 1
        else:
            end = mid - 1
            
    return ans
            
    

File: 프로그래머스/3/64062. 징검다리 건너기/징검다리 건너기.java
import java.util.Arrays;

class Solution {
    public int solution(int[] stones, int k) {
        final int N = stones.length;
        int end = Arrays.stream(stones).max().getAsInt();
        int start = Arrays.stream(stones).min().getAsInt();
        int answer = Integer.MAX_VALUE;

        while (start <= end) {
            int mid = (start + end) / 2;
            int[] tmp = new int[N];
            for (int i = 0; i < N; i++) {
                tmp[i] = stones[i] - mid;
            }
            int cnt = 0;
            boolean possible = true;
            for (int num : tmp) {
                if (num <= 0) cnt++;
                else cnt = 0;
                if (cnt == k) {
                    possible = false;
                    answer = Math.min(answer, mid);
                    break;
                }
            }
            if (possible) {
                start = mid + 1;
            } else {
                end = mid - 1;
            }
        }


        return answer;
    }
}